src/electron/device-connect-view/components/device-connect-port-entry.tsx

```diff
@@ -41,38 +42,35 @@ export class DeviceConnectPortEntry extends React.Component<DeviceConnectPortEnt
                     maskChar=""
                     mask="99999"
                 />
-                <Button
-                    primary={this.props.needsValidation}
-                    disabled={this.state.isValidateButtonDisabled}
-                    className="button-validate-port"
-                    onClick={this.onValidateClick}
-                >
-                    Validate port number
-                </Button>
+                {this.renderValidationPortButton()}
             </div>
         );
     }
 
-    private onPortTextChanged = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
-        this.setState({ isValidateButtonDisabled: !newValue || newValue === '', port: newValue });
-        this.props.updateStateCallback(DeviceConnectState.Default);
-    };
+    private renderValidationPortButton(): JSX.Element {
+        const props = {
+            disabled: this.isValidateButtonDisabled(),
+            className: 'button-validate-port',
+            onClick: this.onValidateClick,
+        };
 
-    private onValidateClick = async (event: React.MouseEvent<HTMLButtonElement>): Promise<void> => {
-        this.setState({ isValidateButtonDisabled: true });
+        if (this.props.viewState.deviceConnectState !== DeviceConnectState.Connected) {
+            return <PrimaryButton {...props}>Validate port number</PrimaryButton>;
+        }
 
-        const port = parseInt(this.state.port, 10);
+        return <DefaultButton {...props}>Validate port number</DefaultButton>;
+    }
 
-        this.props.deps.deviceConnectActionCreator.validatePort(port);
-        this.props.updateStateCallback(DeviceConnectState.Connecting);
+    private isValidateButtonDisabled(): boolean {
+        return !this.state.port || this.state.port === '' || this.props.viewState.deviceConnectState === DeviceConnectState.Connecting;
+    }
 
-        await this.props.deps
-            .fetchScanResults(port)
-            .then(data => {
-                this.props.updateStateCallback(DeviceConnectState.Connected, `${data.deviceName} - ${data.appIdentifier}`);
-            })
-            .catch(err => this.props.updateStateCallback(DeviceConnectState.Error));
+    private onPortTextChanged = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
+        this.setState({ port: newValue });
```

Comment:
what happens if emitChanged event is fired when the user is typing?

Reply:
never mind. it will only change if the props value changes for this object

Reply:
if it helps, react handle state changes as async operations, it tries to merge them before applying the new state values.